### PR TITLE
Use BlockingCollection in SharedTextWriter

### PR DESCRIPTION
**Pull request by mholo65**
Did some profiling and noticed a lot of time was spent on `ShareTextWriter.WriteLine` and `SharedTextWriter.WriteLineAsync`.

Noticed the `lock` and figured we could use `SemaphoreSlim` instead.

### CODE REVIEW

File: src/OmniSharp.Host/Services/SharedTextWriter.cs [hunk 1]
```diff
@@ -19,26 +19,33 @@ public SharedTextWriter(TextWriter writer)
         protected override void DisposeCore(bool disposing)
         {
             // Finish sending queued output to the writer.
-            _task.Wait();
+            _semaphoreSlim.WaitAsync().GetAwaiter().GetResult();
         }
 
         public void WriteLine(object value)
```
Comment (david-driscoll) [inline src/OmniSharp.Host/Services/SharedTextWriter.cs]: Can we refactor this method away?

**Comment by mholo65** [comment]
image

File: src/OmniSharp.Host/Services/SharedTextWriter.cs (hunk 1 above)
Comment (david-driscoll) [reply src/OmniSharp.Host/Services/SharedTextWriter.cs]: looking at the code it looks like it's called in other async contexts.

**Comment by mholo65** [comment]
Could do some benchmarks to see if there's any actual performance boost.

File: src/OmniSharp.Host/Services/SharedTextWriter.cs (hunk 1 above)
Comment (mholo65) [reply src/OmniSharp.Host/Services/SharedTextWriter.cs]: Yeah, it's not used anymore. As we don't ship NuGets, we probably don't care about obsoleting, just remove it from the interface?

**Comment by david-driscoll** [comment]
I just figure if we can avoid blocking a thread at all it would probably be better.

File: src/OmniSharp.Host/Services/SharedTextWriter.cs (hunk 1 above)
Comment (david-driscoll) [reply src/OmniSharp.Host/Services/SharedTextWriter.cs]: yeah, I say just nuke it before we start shipping nugets 😄 👍

**Comment by rchande** [comment]
@mholo65 What do you think of implementing this with BlockingCollection<T>? https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.blockingcollection-1?view=netcore-3.0

**Comment by david-driscoll** [comment]
@rchande I'd be fine with that.

@mholo65 lsp already has this with the output handler https://github.com/OmniSharp/csharp-language-server-protocol/blob/7d8567c09c88d27b4b4a2209959f689df8e9419d/src/JsonRpc/OutputHandler.cs

**Comment by rchande** [comment]
Actually, blockingCollection blocks a thread, so this is probably better.

**Comment by david-driscoll** [comment]
A blocking collection will only block it's own thread, not the calling task, assuming you just queue the item up.  For input / output it probably makes more sense to just have the dedicated thread, then we can queue data up for it as fast as we're allowed, but it'll write out at whatever rate it can.

As long as the order is guaranteed (which by default `BlockingCollection` uses `ConcurrentQueue`) I think dumping handling the writing in another thread might make sense.

Thoughts? @mholo65 @rchande  

As it is I'm fine with this change, but it might almost make more sense to refactor it entirely.

**Comment by mholo65** [comment]
@david-driscoll @rchande I think that BlockingCollection is even better, then we'll dump the IO to another thread, and won't block main thread.

**Comment by mholo65** [comment]
@david-driscoll @rchande @filipw I modified this PR to use `BlockingCollection` instead, also no need for the Async version anymore.

**Comment by filipw** [comment]
thanks 👏

**Pull request merged**